[cisco_asa_failover] add support for firepower devices

thl-cmk opened the pull request:
This is the first followup to PR 350

jherbel commented:
Hi, looks good, you can also include the renaming here, but please provide a good explanation which I can include in a werk.

jherbel commented:
Could you please provide a short explanation why this renaming makes sense (not in the code, simply as a comment here)? I will then include it in a werk.

thl-cmk commented:
The Cisco ASA supports three modes of redundancy / scalability.
 - Failover, the one this check monitors
 - Cluster, what the current service name suggests
 - VPN load balancing cluster, not relevant in this context

Failover consist always of two identical (hardware/software/licenses)
ASA devices. These two devices form a failover pair where one device is
active (handles the Trafic) and the other device is standby. In case of a
failure of the active unit the standby units takes over the active role
and starts to handle the traffic. Both units are connected over a failover link
and optionaly with a state link. In most configuration the failover link and
the state link share one physical interface. In this mode allmost all functions
of the ASA are supported (I belive except for the local CA server). To distinguish
the devices on device is configured as primary on as secondary unit.

In a ASA cluster there can be up to 16 identical devices (ASA5585, Firepower 4100/9330).
The are all active (handle traffic) at the same time. These devices are connected via one control link.
Over this control link the cluster is configured and state information is
passed between the cluster members. In case of asynchron traffic the packets are
routed to the correct device via the cluster link. In this mode of configuration
are many function of the ASA not supported. Failover is particularly important here.

VPN load balancing cluster is mainly for loadbalancing RA VPN sessions. It can be usesd
stand alone or on top of a failover konfiguration (can not be used with cluster mode).
The devices involved hafe not to be identical, there is no state or configuration
synchronisation.

More details can be found in the ASA configuration guide.
https://www.cisco.com/c/en/us/td/docs/security/asa/asa97/configuration/general/asa-97-general-config/ha-cluster.html

This picture gives an overview of the traffic flow in the ASA cluster
https://www.cisco.com/c/dam/en/us/td/i/300001-400000/330001-340000/333001-334000/333480.jpg

jherbel commented:
Ok, looks good, thank you. If you want, you can also incorporate your last point in this PR (since the actual changes are very small from what I can see):

> change the detection of H/A device from "failover off" to "not configured"
> "failover off" means failover is not enabled, this can be because its not
> fully configured, or because of some maintenance state. -> Should trigger a warning
> Whereas "not configured" means, it's no failover device

thl-cmk commented:
Well, at the moment i need to do a little more digging for that. It's not so clear as I thought :-( . At the moment I would only add WATO options to make the failed device/failoverlink state configurable.

jherbel commented:
Thats also ok. Do you want to do this here or in a separate PR (I guess you saw that we cannot accept any new feature PRs at the moment, however, for this, I would make an exception since this is a follow-up to PR #350)?

thl-cmk commented:
>>  Do you want to do this here or in a separate PR...
Your decision. I can do this in here or open a followup PR. 
About this stop for new PRs any idea how long this will last?

jherbel commented:
Then please add it to this PR. Concerning the stop for new PRs, I have no idea, sry. Our problem really is that we are simply not able to keep up with the incoming PRs and all the other work we have (tickets, developments etc).

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -14,142 +15,147 @@
 )
 
 
+def get_cisco_asa_state_name(st: str) -> str:
```
Why did you modify this? Please leave it as it is (defined further below, marked as private, the dict defined outside the function and using an f-string).

thl-cmk replied:
need the dict only in the litle helper function..., changed it back.

> using an f-string
from your guidlines: Use classic format strings (%s) for the time being. We'll move over to the new format() syntax in the future, but for the moment we'd like to stay consistent.

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -14,142 +15,147 @@
+    yield Result(state=State.OK,
+                 summary='Device (%s) is the %s' % (section.local_role, section.local_status_detail))
 
-    if params[section.local_role] != _get_cisco_asa_state_name(section.local_status):
```
Why did you modify this (the formatting looked much better before and f-strings are also better)?

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -14,142 +15,147 @@
 # .1.3.6.1.4.1.9.9.147.1.2.1.1.1.4.7  'Standby unit'
 
+
+# [['Failover LAN Interface', '2', 'LAN_FO GigabitEthernet0/0.777'],
```
remove

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -4,6 +4,7 @@
 # This file is part of Checkmk (https://checkmk.com). It is subject to the terms and
 # conditions defined in the file COPYING, which is part of this source code package.
 
+#
```
remove

jherbel reviewed (changes requested):
Before I review this properly, please re-base on the master. You are introducing unnecessary changes here, see the examples below.

jherbel reviewed (changes requested):
There are still unnecessary changes, such as moving functions around and changing the fetch-argument of register.snmp_section....
Concerning the string formatting, it is true that our guidelines still say that we should use %-formatting, however, we started to switch to f-strings now, so please also use them.

thl-cmk commented:
have done some more rollback. About the fetch-argument of register.snmp_section:  found it, rollback in the next commit :-(

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -40,30 +42,34 @@ class Section:
     failover = {}
     for role, status, detail in string_table:
-        if 'this device' in role and detail.lower() != 'failover off':
+        if 'this device' in role and not detail.lower() == 'failover off':
```
Please roll this back

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -92,36 +98,26 @@ def _get_cisco_asa_state_name(st: str) -> str:
-        summary=f'Device ({section.local_role}) is the {section.local_status_detail}',
-    )
+    yield Result(state=State.OK,
+                 summary=f'Device ({section.local_role}) is the {section.local_status_detail}')
```
Please roll this back

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -92,36 +98,26 @@ def _get_cisco_asa_state_name(st: str) -> str:
-            state=State(params['failover_state']),
-            summary=f'(The {section.local_role} device should be {params[section.local_role]})',
-        )
+    if not params[section.local_role] == _get_cisco_asa_state_name(section.local_status):  # wrong device active/standby
```
Please roll this back

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -92,36 +98,26 @@ def _get_cisco_asa_state_name(st: str) -> str:
-            summary=
-            f'Failover link state is {_get_cisco_asa_state_name(section.failover_link_status)}',
-        )
+    if section.failover_link_status not in ['2']:  # not up
```
Use `!= '2'`

jherbel commented on cmk/base/plugins/agent_based/cisco_asa_failover.py:
```diff
@@ -133,23 +129,26 @@ def check_cisco_asa_failover(params: (Mapping[str, Any]), section: Section) -> C
             '3',  # CISCO-FIREWALL-MIB::cfwHardwareStatusValue
             '4',  # CISCO-FIREWALL-MIB::cfwHardwareStatusDetail
-        ],
+        ]
```
Roll this back

jherbel reviewed (changes requested):
Please also change the formatting of the tests to the way it was before, otherwise, its impossible to spot any actual differences

thl-cmk commented:
I started over from the last version we agreed on. 

Added the WATO options and some minor changes.

- insert the the failover link name in the result for failover link not up
- replace the "if len(failover) == 5:" with a try/except block, so on further changes this line needs not to be changed again
- removed the "{UNKNOWN}" state sentence form checkman, if I don't miss something, the check is not doing this.

Sorry for the mess. Hope we are back on track now.
